### PR TITLE
ENH: add max dist to NearestNDInterpolator

### PR DESCRIPTION
**Pull request by harshilkamdar**
#### Reference issue
This does not close a particular issue as far as I know, but does speed up nearest neighbor interpolation by a siginificant amount in cases where a max distance arg is passed.

#### What does this implement/fix?
This allows for far quicker queries and nearest neighbor interpolation for large datasets where prior information about when/where to interpolate is known. In some 2D geospatial examples with tens of millions of points, this leads to a speedup of anywhere from 2-30x. 

#### Additional information
n/a. This is my first PR here -- please let me know if I did anything wrong.

### CODE REVIEW

File: scipy/interpolate/_ndgriddata.py
Comment (j-bowhay) [inline scipy/interpolate/_ndgriddata.py:97]:
```suggestion
    def __init__(self, x, y, rescale=False, tree_options=None, *, query_max_dist=np.inf):
```

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:51]: New features go to 1.12, I'd expect.

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:125]: Since this new argument is forwarded to `kdtree.query`, maybe better keep the name consistent. Maybe even expose `**kwargs` and forward them all to `self.tree.query`. Not sure if they all make sense for the `NearestNDInterpolator`, so maybe start with `distance_upper_bound` and `workers`. Not sure about the rest, WDYT?

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:127]: Is this what `kdtree` returns? Would be great to stress it in the comment if it's indeed the case.

**Review by ev-br — CHANGES_REQUESTED** [review]
Looks very reasonable!

My main comment is about naming: I think it would be better to 
1) keep the name of the parameter consistent with KDTree, and 
2) maybe think about exposing other `KDTree.query` parameters where it makes sense. 

The second part here is optional of course.

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py]: Good catch on also exposing workers. The rest don't seem useful. Having **kwargs might be a little dangerous because the interpolator will break if k is not set to 1. I have exposed workers and distance_upper_bound (renamed) separately in the latest commit: https://github.com/scipy/scipy/pull/19483/commits/2fa01c9110a7c4db730625ac2efb57e94b8d7f3d. Let me know what you think

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py]: Added some more content in the text to stress this point and clarify.

**Comment by ev-br** [comment]
One other thing (sorry for the piecemeal review, only noticed this being next to `**tree_options`): you are adding these new arguments to the constructor and only use them at the call site. Is there a need to store `workers` and `distance_upper_bound` on the instance? Maybe better mirror KDTree and only supply them in `__call__`.

**Comment by harshilkamdar** [comment]
No worries at all. My rationale was looking to [`interpolate.griddata`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.griddata.html#scipy.interpolate.griddata) in the future and making it more flexible. 

Right now, griddata does not support passing any options to the underlying interpolator (nearest, linear, or cubic). I'm hoping to change that in a new PR and figured it would be cleaner to only expose and use class-level options and keeping __call__ clean for now. Linear doesn't have interesting options, but CloughTocher2DInterpolator [does](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CloughTocher2DInterpolator.html#scipy.interpolate.CloughTocher2DInterpolator). 

This is not a super strong reason, so I'm happy to go with whatever you prefer.

**Comment by ev-br** [comment]
Right. So delineating between `__init__` and `__call__` optional arguments would require two sets of dicts as inputs to `griddata` (unpacked or not), and those would be `method`-dependent. Not nice. 
OTOH, you current approach is that everything is passed in a constructor and `__call__` signature only has data points and nothing else. Which is a reasonable invariant, so let's roll with it I'd say. WDYT @j-bowhay ?

I'd then ask for a small tweak:
- please add a leading underscore to things you store on the instance (as in `self._workers` is private, while `self.workers` is a part of public API).

In fact, it's not immediately clear from the docstring that this whole interpolator is for unstructured data and is KDTree-based. Would be great if you could update that in a follow-up PR.

**Comment by j-bowhay** [comment]
I appreciate the argument around future work on `griddata` but it seems cumbersome that I would have to rebuild my interpolator just because I want to change the workers / max dist. Could we not add an `options` dict argument to `griddata` (as we do for minimise and friends) then each method uses this where needed (whether that be `__init__` or `__call__`)

**Comment by harshilkamdar** [comment]
@j-bowhay, @ev-br: thank you both for your careful feedback. I have updated the PR based on both your comments in the latest commit - https://github.com/scipy/scipy/pull/19483/commits/ff1d52ffe1a827bf4941ccbc213dd5c21939f767. It now takes in a `query_options` dict and exposes some more cKDTree functionality. I've also added a couple more tests.

> In fact, it's not immediately clear from the docstring that this whole interpolator is for unstructured data and is KDTree-based. Would be great if you could update that in a follow-up PR.

I will do this in a separate documentation PR.

**Comment by j-bowhay** [comment]
Sorry I wasn't quite clear, I was proposing the `griddata` take some kind of option dict to pass to the underlying interpolator and a signature like `NearestNDInterpolator.__call__(*args, distance_upper_bound=np.inf, etc)` but perhaps lets see what @ev-br thinks so we don't have to make lots of back and forth changes

**Comment by ev-br** [comment]
To be clear: I'm fine with either approach. I'm sold on that the inconvenience of lumping together constructor and call arguments is rather minor and a cleaner `griddata` signature outweights it; I can also live with a `optimize`-like approach. As long as we are not adding public attributes to interpolators, I'm OK and defer to @j-bowhay for the final API specification.

**Comment by harshilkamdar** [comment]
Sounds good - sorry about the added confusion. @j-bowhay let me know which option you prefer and I can implement. Either is fine on my end and should likely be minimal effort

File: scipy/interpolate/_ndgriddata.py
Comment (j-bowhay) [inline scipy/interpolate/_ndgriddata.py:101]:
```suggestion
    def __call__(self, *args, **query_options):
```

File: scipy/interpolate/_ndgriddata.py
Comment (j-bowhay) [inline scipy/interpolate/_ndgriddata.py:129]:
```suggestion

```

File: scipy/interpolate/_ndgriddata.py
Comment (j-bowhay) [inline scipy/interpolate/_ndgriddata.py:136]:
```suggestion
        dist, i = self.tree.query(xi, **query_options)
```

**Review by j-bowhay — COMMENTED** [review]
How about an interface like this?

**Comment by harshilkamdar** [comment]
Thanks both for your patience. @j-bowhay - I like it. I have committed and updated tests + docs and have re-requested review.

**Comment by ev-br** [comment]
CI failures look related, could you please take a look

**Comment by harshilkamdar** [comment]
Sorry, dumb bug. Fixed now.

**Comment by ev-br** [comment]
CI is still unhappy :-(.
Might be faster to iterate locally: the incantation is `$ python dev.py test -s interpolate` or `-t path/to/test/file`.

**Comment by harshilkamdar** [comment]
@ev-br - good call. 

Through the `griddata` test failures, I've found that both `NearestNDInterpolator` and `griddata` are silently actually supporting two things that I wasn't originally accounting for - the case where `__call__` gets an n-D `xi` argument and the case where the y-values for the interpolation are also multidimensional. 

I've come up with hopefully an acceptable solution to account for these two cases and updated docs & comments for NearestNDInterpolator to reflect this. I couldn't get scipy to build locally for some annoying reasons but have tried to make sure that all tests in `test_ndgriddata.py` pass by checking manually. If there end up being failures, I can fix them quickly.

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:35]: OK, so the original was papering over a detail that `y` can have trailing dimensions (BTW, I'm not sure LinearNDInterpolator and CT can). 
The new version is plain confusing, my mind starts racing to figure out what is `n` in  `n-D`. Essentially, what is going on here is that `y` is a 1D array of data values or a stack of 1D arrays of data values, correct? I'd probably say something along these lines then (ML people like to talk about batch dimensions).

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:33]: `x` is surely a 2D array, no trailing dimensions allowed.

**Review by ev-br — COMMENTED** [review]
OK, this is getting close! I like the flatten-compute-restore logic, it makes a lot of sense. 

I've left several comments about 1) minor simplifications of implementation details and 2) clarity of comments/documentation.

The story about stacks of arrays is always confusing, so let's try to clarify it here. So, there are three arrays: `x`, `y` and `xi`. 

Let's see if I got it right:

- `x` is always 2D
- `y` is 1D or a stack of 1D, where the first dimension is the same as the second (or first?) dimension of `x`, and trailing dimensions represent the stack of 1D arrays.
- `xi` is normally a 1D array of dimension `ndim` or a stack of 1D arrays, where the stacking is along the _leading_ dimension (this is what _broadcastable_ word hints at inhttps://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.NearestNDInterpolator.__call__.html#scipy.interpolate.NearestNDInterpolator.__call__ ).

I think it would be best to explain this (or a correct version of it) to clearly separate what is the number of dimensions, what is the number of data points and how stacked dimensions come into play.
Two general considerations to keep in mind:

- the docs for NearestNDInterpolator, LinearNDInterpolator and CloughtTocher2DInterpolator should look relatively consistent. If one of them supports trailing/leading dimensions and the other does not, OK, so be it. But going from nearest to linear interpolator in a simple case should be seamless, and so should be the docs.
- Maybe it'd be useful to take a look at how RegularGridInterpolator documents essentially the same issue. Ideally, regular and scattered interpolants look similar.

Finally, since the documentation story is not strictly speaking related to your original changes, it would be perfectly fine to undo the docstring changes and keep the documentation update for a follow-up PR (which would be very welcome indeed).

File: scipy/interpolate/_ndgriddata.py
Comment (j-bowhay) [inline scipy/interpolate/_ndgriddata.py:111]: This 
```suggestion
        **query_options
```

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py:154]: To clarify - is what I've wrriten right now adequate or should I modify?

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py:149]: Exactly. There were a few tests where self.values were integers and it caused numpy weirdness when I tried to assign nan's to an array declared as int. example below 


```
In [1]: arr = np.full(5, np.nan, dtype=np.int64)

In [2]: arr
Out[2]: 
array([-9223372036854775808, -9223372036854775808, -9223372036854775808,
       -9223372036854775808, -9223372036854775808])

```

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py:129]: Yeah, I'm not being careful here in my words. Have given the description another go.

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py:125]: Yeah, this is confusing and griddata is not relevant - updated.

**Comment by harshilkamdar** [comment]
@ev-br - thanks for the thorough comments. I think you're right that I'm confusing nD/2D and trailing dimensions. The pointer to RegularGridInterpolator was very helpful.

I have considered most of your comments and hopefully addressed bulk of it in the new commit. Hopefully this is a more accurate version of what's going on.

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [reply scipy/interpolate/_ndgriddata.py:155]: I would be tempted to just write `interp_values[valid_mask] = self.values[i[valid_mask], ...]` but it's OK, can do in a sequel. No need to push a fix just for this one.

Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py:155]: Suonds good - will do

File: scipy/interpolate/_ndgriddata.py
Comment (harshilkamdar) [reply scipy/interpolate/_ndgriddata.py]: I realized my response to this didn't go through. Yeah, you're right - I've fixed this in latest commit. Looks like Linear and CT can actually. There's some voodoo going on in NDInterpolatorBase's _set_values, where it'll recast values to be 2D (npoints, nvalues) for all interoplators that subclass.

File: scipy/interpolate/_ndgriddata.py
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:116]:
```suggestion
            .. versionadded:: 1.12.0
```

**Review by ev-br — APPROVED** [review]
All my comments have been addressed, I think it is a great addition, so +1 from me. 

Unless there are further comments, am going to hit the green button later this week.

**Review by j-bowhay — APPROVED** [review]

**Comment by ev-br** [comment]
Am going to merge now, based on two core dev approvals. Thanks Jake for the review, thank you @harshilkamdar for the enhancement and congrats with what I believe is your first SciPy commit. Keep them coming!

**Pull request merged**
